src/lib.rs

```diff
@@ -1,287 +1,717 @@
+
+        let res = unsafe {
+            cubeb_stream_init(ctx.native, &mut stm, stream_name.as_ptr(),
+                transmute(in_device), transmute(in_params.as_ref()),
+                transmute(out_device), transmute(out_params.as_ref()),
+                latency_frames, Some(T::data_cb_ffi()), nat_state_cb,
+                &mut *data as *mut StreamData<T> as *mut c_void
+            )
+        };
+        match res {
+            CUBEB_OK => Ok(Stream {
+                native: stm,
+                data: data,
+            }),
+            _ => Err( Error::from(res) )
+        }
+    }
+
+    pub fn start(&self) -> Result<()> {
+        let res = unsafe {
+            cubeb_stream_start(self.native)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn stop(&self) -> Result<()> {
+        let res = unsafe {
+            cubeb_stream_stop(self.native)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn reset_default_device(&self) -> Result<()> {
+        let res = unsafe {
+            cubeb_stream_reset_default_device(self.native)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn position(&self) -> Result<u64> {
+        let mut val = 0;
+        let res = unsafe {
+            cubeb_stream_get_position(self.native, &mut val)
+        };
+        match res {
+            CUBEB_OK => Ok(val),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn latency(&self) -> Result<u32> {
+        let mut val = 0;
+        let res = unsafe {
+            cubeb_stream_get_latency(self.native, &mut val)
+        };
+        match res {
+            CUBEB_OK => Ok(val),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn set_volume(&self, volume: f32) -> Result<()> {
+        let res = unsafe {
+            cubeb_stream_set_volume(self.native, volume)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+
+    pub fn set_panning(&self, panning: f32) -> Result<()> {
+        let res = unsafe {
+            cubeb_stream_set_panning(self.native, panning)
+        };
+        match res {
+            CUBEB_OK => Ok(()),
+            _ => Err(Error::from(res)),
+        }
+    }
+}
+
+impl<T: Sample> Drop for Stream<T> {
+    fn drop(&mut self) {
+        unsafe { cubeb_stream_destroy(self.native); }
+    }
+}
+
+
+pub fn print_state_change(state: State) {
+    match state {
+        State::Started => println!("stream started"),
+        State::Stopped => println!("stream stopped"),
+        State::Drained => println!("stream drained"),
+        State::Error => println!("stream error"),
+    }
+
+}
+
+
+extern fn data_callback_i16(_stm: *const cubeb_stream,
```

Comment:
The function *itself* should be unsafe, you can trivially pass null or invalid pointers.